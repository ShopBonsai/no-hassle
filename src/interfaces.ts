import * as Joi from '@hapi/joi';
import { RequestHandler } from 'express';
import { HttpMethod } from './constants';

export interface IRouteResult {
  get: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  post: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  put: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  patch: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  delete: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  template: (template: ITemplateRoute, ...args: RequestHandler[]) => IRouteResult;
}

export enum SchemeType {
  Http = 'http',
  Https = 'https',
  Ws = 'ws',
  Wss = 'wss',
}

export interface ISwaggerBaseDefinition {
  title: string;
  description: string;
  host: string;
  schemes: SchemeType[];
  basePath: string;
  contact?: { email: string };
  version?: string;
  apiVersion?: string;
}

export interface ISwaggerDefinition {
  swagger: string;
  host: string;
  basePath: string;
  schemes: SchemeType[];
  info: {
    description: string;
    version: string;
    title: string;
    contact: {
      email: string;
    };
  };
  paths: {};
  definitions: {};
}

export enum AuthenticationTypes {
  BASIC = 'basic',
  API_KEY = 'apiKey',
}

export type AuthenticationType = `${AuthenticationTypes}`;

export type AuthenticationOptions = { apiKeyHeaderName?: string };

export interface ISwaggerOptions {
  title?: string;
  description?: string;
  host?: string;
  schemes?: SchemeType[];
  basePath?: string;
  apiVersion?: '2.0' | '3.0';
  version?: string;
  auth?: AuthenticationType;
  authOptions?: AuthenticationOptions;
}

export interface IInput {
  body?: ISchema;
  query?: ISchema | object;
  params?: ISchema | object;
  headers?: ISchema | object;
}

export type IOutput = {
  [key: number]: ISchema | { [key: string]: ISchema };
};

export interface IDocsOptions {
  input?: IInput;
  output?: IOutput;
  contentTypes?: string[];
  schemaOptions?: object;
  tags?: string[];
  description?: string;
  summary?: string;
  /**
   * The prefix to add to the route path.
   */
  prefix?: string;
  /**
   * Whether to omit the URL prefix from the generated documentation.
   */
  shouldOmitPrefixInDocs?: boolean;
}

export interface ITemplateRoute extends IDocsOptions {
  method: HttpMethod;
  path: string;
}

export interface IExecOptions extends IDocsOptions {
  /**
   * Whether to generate documentation for this route.
   */
  docs?: boolean;
  /**
   * The HTTP method of the route.
   */
  method: HttpMethod;
  /**
   * The path of the route.
   */
  path: string;
}

export interface IOptions extends IDocsOptions {
  docs?: boolean;
}

export interface IValidateResponse {
  isValid: boolean;
  errors?: Joi.ValidationErrorItem[];
}

export type ISchema = Joi.Schema & {
  _meta?: [];
};
