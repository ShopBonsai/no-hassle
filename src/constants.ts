export enum HttpMethod {
  Get = 'get',
  Post = 'post',
  Put = 'put',
  Patch = 'patch',
  Delete = 'delete',
}
