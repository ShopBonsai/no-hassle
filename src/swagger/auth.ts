enum AuthenticationType {
  BASIC = 'basic',
}

const basicConfig = {
  securityDefinitions: {
    basicAuth: {
      type: 'basic',
    },
  },
  security: [{ basicAuth: [] }],
};

const authenticationConfiguration = {
  [AuthenticationType.BASIC]: basicConfig,
};

export const getAuthentication = (type?: 'basic') => authenticationConfiguration[type];
