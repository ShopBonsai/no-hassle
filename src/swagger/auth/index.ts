
import { Authentication, Security } from '../../interfaces';
import { IOutputAuthentication, OutputSecurity } from './interfaces';
import { getSecurityDefinitions, validateSecurity } from './util';

/**
 * Generates the security object that can be used in the swagger spec.
 * Can be used for both default security and operation level security.
 * Does not support scoping for OAuth2 security definitions.
 * @param {Security} security - Ex. [['BasicAuth']]
 * @returns {OutputSecurity} - Ex. [{ BasicAuth: [] }]
 */
export const getSecurity = (security: Security): OutputSecurity => {
  return security.map((sec) => {
    return sec.reduce((acc, securityDefinitionId) => {
      return {
        ...acc,
        [securityDefinitionId]: [],
      };
    }, {});
  });
};

/**
 * Converts simplified authentication object to the format expected by the swagger spec
 * for the securityDefinitions and security properties.
 * @param {Authentication} auth - Simplified authentication object, if not provided, will return an empty object.
 * @returns {IOutputAuthentication} - Authentication object that can be used in the swagger spec.
 */
export const getAuthentication = (auth?: Authentication): IOutputAuthentication => {
  const { securityDefinitions, security = [] } = auth || {};
  if (!securityDefinitions) {
    return {};
  }

  validateSecurity(security, securityDefinitions.map(({ id }) => id));

  return {
    securityDefinitions: getSecurityDefinitions(securityDefinitions),
    security: getSecurity(security),
  };
};
