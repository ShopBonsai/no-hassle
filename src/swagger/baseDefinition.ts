const j2s = require('joi-to-swagger');

import { ISwaggerDefinition, ISwaggerBaseDefinition } from '../interfaces';
import {
  unknownErrorSchema,
  badRequestErrorSchema,
  notFoundErrorSchema,
  unauthorizedErrorSchema,
} from './errors';

export const baseDefinition = ({
  title,
  description,
  host,
  schemes,
  basePath,
  contact,
  version,
  apiVersion = '2.0',
}: ISwaggerBaseDefinition): ISwaggerDefinition => ({
  host,
  schemes,
  basePath,
  swagger: apiVersion,
  info: {
    title,
    description,
    version,
    contact,
  },
  paths: {},
  definitions: {
    BadRequestError: j2s(badRequestErrorSchema).swagger,
    NotFoundError: j2s(notFoundErrorSchema).swagger,
    UnauthorizedError: j2s(unauthorizedErrorSchema).swagger,
    UnknownError: j2s(unknownErrorSchema).swagger,
  },
});
