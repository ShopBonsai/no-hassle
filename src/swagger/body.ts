import { IInput, ISwaggerDefinition } from '../interfaces';
import { getDefinition } from './definition';

export const getBody = (swagger: ISwaggerDefinition, input?: IInput) => {
  if (!input) return [];

  const { body } = input;

  if (body) {
    const definition = getDefinition(swagger, body, 'Input');
    return {
      content: {
        'application/json': {
          schema: {
            $ref: `#/components/schemas/${definition}`,
          },
        },
      },
      required: true,
    };
  }
};
