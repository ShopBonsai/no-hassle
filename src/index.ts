import { Application, RequestHandler } from 'express';
import { isCelebrate } from 'celebrate';
import { getSwagger, generateSwagger } from './swagger';
import { validateSchema, validateValue } from './lib/validator';
import { IOptions, IRouteResult, IExecOptions, ITemplateRoute } from './interfaces';
import { HttpMethod } from './constants';

const addRoute = (
  app: Application,
  options: IExecOptions,
  ...args: RequestHandler[]
): IRouteResult => {
  const {
    method,
    path: rawPath,
    prefix = '',
    input = null,
    schemaOptions = {},
    docs = true,
  } = options;
  const path = rawPath === '/' ? prefix : `${prefix}${rawPath}`;

  // TODO: Add method overloading -> no options provided -> basic result

  // Autogenerate documentation based on the schema
  if (docs) {
    generateSwagger(path, method, options);
  }

  // Add express route (with optional validation)
  input && Object.keys(input).length > 0
    ? app[method](path, validateSchema(input, schemaOptions), ...args)
    : app[method](path, args);

  // Return function instance with same router for currying
  return router.use(app, prefix);
};

export const router = {
  use: (app: Application, prefix: string = ''): IRouteResult => ({
    get: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Get }, ...args),
    post: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Post }, ...args),
    put: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Put }, ...args),
    patch: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Patch }, ...args),
    delete: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Delete }, ...args),
    template: ({ path, method, ...otherOpts }, ...args) =>
      addRoute(app, { ...otherOpts, path, prefix, method }, ...args),
  }),
};

export const isValidationError = (error: object): boolean => isCelebrate(error);

export * from './swagger/errors';
export * from './constants';
export * from './interfaces';
export { IOptions, ITemplateRoute, getSwagger, validateValue, validateSchema };
