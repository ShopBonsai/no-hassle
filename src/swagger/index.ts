import { IDocsOptions, ISwaggerOptions, ISwaggerDefinition } from '../interfaces';
import { cleanPath } from '../lib/utils';
import { baseDefinition } from './baseDefinition';
import { getParameters } from './parameter';
import { getResponses } from './response';
import { getAuthentication } from './auth';
import { HttpMethod } from '../constants';

enum SwaggerDefaultConfig {
  VERSION = '2.0',
  HOST = '',
}

// TODO: move expectation to consumer
const globalSwagger: ISwaggerDefinition = {
  ...baseDefinition({
    title: 'Default swagger title',
    description: 'The API is documented here',
    host: 'localhost:3000',
    basePath: '/',
    version: '1.0',
    apiVersion: '2.0',
    contact: { email: 'developers@shopbonsai.com' },
  }),
};

export const updateSwagger = (key: string, values: object) =>
  Object.assign(globalSwagger[key], values);

export const generateSwagger = (path: string, method: HttpMethod, options: IDocsOptions) => {
  const {
    input,
    output,
    contentTypes = ['application/json'],
    summary,
    description,
    tags = [],
  } = options;

  const cleanedPath = cleanPath(path);

  const result = {
    [method]: {
      tags,
      produces: contentTypes,
      parameters: getParameters(globalSwagger, input),
      responses: getResponses(globalSwagger, output),
      // optional
      ...(summary && { summary }),
      ...(description && { description }),
    },
  };
  // If path already exists (other method for example)
  if (globalSwagger.paths.hasOwnProperty(cleanedPath)) {
    Object.assign(globalSwagger.paths[cleanedPath], result);
  } else {
    updateSwagger('paths', { [cleanedPath]: result });
  }
};

export const getSwagger = (options: ISwaggerOptions = {}): ISwaggerDefinition => {
  const {
    apiVersion = SwaggerDefaultConfig.VERSION,
    host = SwaggerDefaultConfig.HOST,
    auth,
    ...otherOptions
  } = options;

  const authentication = getAuthentication(auth);
  const result: ISwaggerDefinition = {
    ...globalSwagger,
    ...authentication,
    host,
    info: {
      ...globalSwagger.info,
      ...otherOptions,
    },
  };

  return result;
};
