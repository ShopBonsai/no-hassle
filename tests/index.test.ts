import * as express from 'express';
import * as Joi from '@hapi/joi';

import { getSwagger } from '../src/swagger';
import { router } from '../src/index';

describe('router', () => {
  it('should add a route and generate swagger definition', () => {
    const app = express();

    router.use(app).get(
      '/hello',
      {
        description: 'Hello Swagger',
        summary: 'Hello Swagger',
        tags: ['Hello Swagger'],
        output: {
          [200]: Joi.object({ test: Joi.string().example('test') }).meta({ definition: 'test' }),
        },
      },
      () => {},
    );

    expect(getSwagger()).toMatchSnapshot();
  });
});
