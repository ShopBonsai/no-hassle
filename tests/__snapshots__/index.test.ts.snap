// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`router should add a route and generate swagger definition 1`] = `
{
  "basePath": "/",
  "definitions": {
    "BadRequestError": {
      "description": "Provided input is invalid",
      "properties": {
        "errors": {
          "items": {
            "example": {
              "detail": "",
              "status": 400,
              "title": "Bad input request",
            },
            "properties": {
              "detail": {
                "example": "Fatal issue",
                "nullable": true,
                "type": "string",
              },
              "status": {
                "example": 500,
                "format": "float",
                "type": "number",
              },
              "title": {
                "example": "Unknown error",
                "type": "string",
              },
            },
            "required": [
              "status",
              "detail",
              "title",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "NotFoundError": {
      "description": "Not Found",
      "properties": {
        "errors": {
          "items": {
            "example": {
              "detail": "",
              "status": 404,
              "title": "Not Found",
            },
            "properties": {
              "detail": {
                "example": "Fatal issue",
                "nullable": true,
                "type": "string",
              },
              "status": {
                "example": 500,
                "format": "float",
                "type": "number",
              },
              "title": {
                "example": "Unknown error",
                "type": "string",
              },
            },
            "required": [
              "status",
              "detail",
              "title",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "UnauthorizedError": {
      "description": "Authentication information is missing or invalid",
      "properties": {
        "errors": {
          "items": {
            "example": {
              "detail": "",
              "status": 401,
              "title": "Authentication information is missing or invalid",
            },
            "properties": {
              "detail": {
                "example": "Fatal issue",
                "nullable": true,
                "type": "string",
              },
              "status": {
                "example": 500,
                "format": "float",
                "type": "number",
              },
              "title": {
                "example": "Unknown error",
                "type": "string",
              },
            },
            "required": [
              "status",
              "detail",
              "title",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "UnknownError": {
      "description": "Unknown error",
      "properties": {
        "errors": {
          "items": {
            "example": {
              "detail": "",
              "status": 500,
              "title": "Unknown error",
            },
            "properties": {
              "detail": {
                "example": "Fatal issue",
                "nullable": true,
                "type": "string",
              },
              "status": {
                "example": 500,
                "format": "float",
                "type": "number",
              },
              "title": {
                "example": "Unknown error",
                "type": "string",
              },
            },
            "required": [
              "status",
              "detail",
              "title",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "testResult": {
      "properties": {
        "test": {
          "example": "test",
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "host": "",
  "info": {
    "contact": {
      "email": "developers@shopbonsai.ca",
    },
    "description": "The API is documented here",
    "title": "Default swagger title",
    "version": "1.0",
  },
  "paths": {
    "/hello": {
      "get": {
        "description": "Hello Swagger",
        "parameters": [],
        "produces": [
          "application/json",
        ],
        "responses": {
          "200": {
            "description": "Success",
            "schema": {
              "$ref": "#/definitions/testResult",
            },
          },
          "400": {
            "description": "400 - Bad request",
            "schema": {
              "$ref": "#/definitions/BadRequestError",
            },
          },
          "401": {
            "description": "401 - Unauthorized",
            "schema": {
              "$ref": "#/definitions/UnauthorizedError",
            },
          },
          "404": {
            "description": "404 - Not found",
            "schema": {
              "$ref": "#/definitions/NotFoundError",
            },
          },
          "5XX": {
            "description": "500 - Unknown error",
            "schema": {
              "$ref": "#/definitions/UnknownError",
            },
          },
        },
        "summary": "Hello Swagger",
        "tags": [
          "Hello Swagger",
        ],
      },
    },
  },
  "schemes": [
    "http",
  ],
  "swagger": "2.0",
}
`;
